### PR TITLE
[ConceptViewer] Automated adding of concepts for librairies

### PR DESCRIPTION
**Pull request by NicolasWent**
The algorithm for automating QuickPi concepts (and other lib concepts).

It research for all the available functions the concept which contain this function and add the concept to the Concept viewer.

### CODE REVIEW

File: pemFioi/conceptViewer-1.0-mobileFirst.js
Comment (mblockelet) [inline pemFioi/conceptViewer-1.0-mobileFirst.js]: Try to avoid adding unnecessary spaces ; and more generally speaking, when you commit, check you're not modifying lines you didn't intend to.

File: pemFioi/conceptViewer-1.0-mobileFirst.js
Comment (mblockelet) [inline pemFioi/conceptViewer-1.0-mobileFirst.js]: I have a few issues with that function `cleanAllConcept` :

* It seems pretty unnecessary to have a separate function. It is called only once, for one specific reason (filter out concepts which don't have a python attribute), and has low reusability, as I cannot really think of another scenario where we would want to filter out concepts without a python attribute. The whole function can be replaced by just one line in the function `findConceptByFunction`, which is the only function currently using it.
* The name is misleading : for instance, "filter" would be a better word than "clean", as doesn't really clean anything but just filters out some concepts from the list.
* The description is a bit misleading too : the concepts aren't "useless", they might be useless for a specific purpose, but not in general. You should explain why you call them useless, for instance "concepts which are useless for the function getConceptsFromBlocks"

I think you should just delete this function, and replace it with one line in findConceptByFunction.

File: pemFioi/conceptViewer-1.0-mobileFirst.js
Comment (mblockelet) [inline pemFioi/conceptViewer-1.0-mobileFirst.js]: includedCategories should be renamed as it's not categories anymore (includedConceptIds for instance).

File: pemFioi/conceptViewer-1.0-mobileFirst.js
Comment (mblockelet) [inline pemFioi/conceptViewer-1.0-mobileFirst.js:453]: Note that concepts is only defined line 423, if includeBlocks.standardBlocks is defined ; that might be an error in the previous version, but that's a good moment to fix it.

File: pemFioi/conceptViewer-1.0-mobileFirst.js
Comment (mblockelet) [inline pemFioi/conceptViewer-1.0-mobileFirst.js]: The variable names `concept` and `conceptFunctionName` are misleading : the variable concept will not contain a concept itself, just the list index of that concept in cleanedConcepts. A variable name such as "conceptIdx" would be more accurate.

**Review by mblockelet — CHANGES_REQUESTED** [review]
Aside from the comments, the code works, so it will be good to merge once the feedback is addressed.

File: pemFioi/conceptViewer-1.0-mobileFirst.js
Comment (NicolasWent) [reply pemFioi/conceptViewer-1.0-mobileFirst.js]: I thought it was more clean with the space there.

But I see, if we didn't wrote it in one line, there would not be space there.

I will check with diff before commits to make sure there is no problems this way

File: pemFioi/conceptViewer-1.0-mobileFirst.js
Comment (NicolasWent) [reply pemFioi/conceptViewer-1.0-mobileFirst.js]: For the first point:
I usually do a lots of functions because I find it cleaner because the final code looks like it is written in english.

But I agree that this function is useless since it can be easily replaced by a simple javascript "filter" which takes only one lign.

I also changed the name of the variable to "filteredConcepts" instead of "cleanedConcepts".

File: pemFioi/conceptViewer-1.0-mobileFirst.js
Comment (NicolasWent) [reply pemFioi/conceptViewer-1.0-mobileFirst.js]: Yes, "includedConceptIds" was the default name but "includedConceptIds" make more sense

File: pemFioi/conceptViewer-1.0-mobileFirst.js
Comment (NicolasWent) [reply pemFioi/conceptViewer-1.0-mobileFirst.js:453]: I have put the concept var outside of the "if" for the next version

File: pemFioi/conceptViewer-1.0-mobileFirst.js
Comment (NicolasWent) [reply pemFioi/conceptViewer-1.0-mobileFirst.js]: Yes you are right, thank you.

I thought it was the same behavior as java or other language when you do for (var ... in ...), but it only gets an id. I will think about that next time

File: pemFioi/conceptViewer-1.0-mobileFirst.js
Comment (mblockelet) [inline pemFioi/conceptViewer-1.0-mobileFirst.js]: The arrow notation [is not compatible with older browsers](https://caniuse.com/?search=arrow). We need to keep compatibility with older browsers as reasonably possible.

**Review by mblockelet — CHANGES_REQUESTED** [review]
Thank you for the modifications, just a last one to make.

File: pemFioi/conceptViewer-1.0-mobileFirst.js
Comment (NicolasWent) [reply pemFioi/conceptViewer-1.0-mobileFirst.js]: Ok, I will replace => by the "function" syntaxe, and also I found the post: https://stackoverflow.com/questions/11147662/filter-an-array-in-javascript-cross-browser
Should I add the filter prototype as they wrote if not present?

Comment (mblockelet) [reply pemFioi/conceptViewer-1.0-mobileFirst.js]: No need to, [Array.filter is supported by IE9](https://caniuse.com/mdn-javascript_builtins_array_filter), and as Mathias said in another discussion, we're fine as we don't have to support older than IE9.

File: pemFioi/conceptViewer-1.0-mobileFirst.js
Comment (mblockelet) [inline pemFioi/conceptViewer-1.0-mobileFirst.js]: Missing semicolon at the end of the return.

Comment (NicolasWent) [reply pemFioi/conceptViewer-1.0-mobileFirst.js]: Sorry, have added it in next commit

**Review by mblockelet — APPROVED** [review]
All good for me, thanks for all the modifications.

**Review by smadbe — APPROVED** [review]
No need to wait for my review when Michel has approved 👍

**Pull request merged**
